share/doc/revng/references/abi-definition.md

```diff
@@ -0,0 +1,213 @@
+
+This allows a non-position-based ABI to take advantage of one of the main traits of position-based ABI: ability to use pointers to copy to pass big aggregates.
+
+This is one of the defining features of arm64 ABI, which allows it to get the best of both approaches - an ability to pass aggregates by copy in registers from the non-position-based approach while also being able to sometimes avoid a copy of a bigger aggregate from the position-based approach.
+
+## `NoRegisterArgumentsCanComeAfterStackOnes` (`bool`)
+
+States whether ABI allows a stack argument (mainly if it's too big to be placed in the registers) to precede other register arguments.
+
+For example, if there is a `Big` struct that has to use the stack, and
+a function like
+```cpp
+void function(Big, uint32_t);
+```
+if this value is set to true, both argument will be passed on stack, otherwise, only the struct will.
+
+## `AllowUnnaturallyAlignedTypesInRegisters` (`bool`)
+
+States whether ABI allows unnaturally aligned types[^1] to be passed in registers. When set to false, all the unnatural types will only be put on stack.
+
+## `CalleeIsResponsibleForStackCleanup` (`bool`)
+
+Specifies who is responsible for cleaning the stack after the function call. If equal to `true`, it's the callee, otherwise it the caller.
+
+## `FloatsUseGPRs` (`bool`)
+
+Setting this to true disables usage of vector registers entirely.
+
+## `StackArgumentsUseRegularStructAlignmentRules` (`bool`)
+
+Setting this to true allows stack arguments to use slots smaller than a word (makes them behave like a single struct instead of a bunch of arguments): so they don't get extended unnecessarily. A notable ABI using this is apple flavor of the aarch64.
+
+For example
+```cpp
+void function(/* enough arguments of all the registers */, uint8_t, uint8_t);
+```
+When this is set to true, both `uint8_t` will get placed as if they were a struct, that is without any padding in-between. When this is false, both of them will get extended to register size before getting placed, leading to padding.
+
+## `UseStrictAggregateAlignmentRules` (`bool`)
+
+Treat all the structs passed using the stack as if they were aligned unnaturally[^1].
+
+This means that they are going to handled with extreme care (which usually just means extra padding).
+
+## `StackAlignment` (`unsigned`)
+
+States the required alignment of the stack at the point of a function call in bytes.
+
+> **_Note:_** states minimum value for ABIs supporting multiple different stack alignment values, for example, if the ABI requires the stack to be aligned on 4 bytes for internal calls but on 8 bytes for interfaces (like 32-bit ARM ABI), the value of `StackAlignment` should be equal to 4.
+
+
+## `MinimumStackArgumentSize` (`unsigned`)
+
+States the minimum possible stack argument size in bytes. When the actual type of an argument is below this value, some padding with be inserted.
+
+## `StackBytesAllocatedForRegisterArguments` (`unsigned`)
```

Comment:
AFAIU these are bytes *before* or *after* the stack arguments?

```
lol(1,2,3);
```
Is it
```
push 1
push 2
push 3
push --
push --
push --
call
```
or
```
push --
push --
push --
push 1
push 2
push 3
call
```

Reply:
They are the first four arguments. All the ABIs we support that use these provide as many as there are register arguments. As far as I understand, they were originally untended to mirror register values, but that is not guaranteed anymore.

Reply:
1. OK, so these are "shadow arguments" and where they are on the stack depends on whether arguments are pushed LTR or RTL. Add a note about this.
2. Now that I think about it, we don't have an option to specify if stack arguments are pushed LTR or RTL. Shouldn't we have such a things?

Reply:
0. Ok, I was wrong about the intention. Digging for the pointers-to-copy, I also ran into this.

They are intended to just "save" volatile registers, but more efficiently (?), since you don't need to move stack pointer - you already have space ready for them.

As far as I understand there are no limitation on how that extra space is used though.

1. 👌.

2. I think we had it at some point, but you didn't like the fact that it was unused, so we got rid of it, or something like that.

Reply:
> I think we had it at some point, but you didn't like the fact that it was unused, so we got rid of it, or something like that.

Mmmh, we support both x86 `cdecl` and `PASCAL` which have different ways to push arguments: https://en.wikipedia.org/wiki/X86_calling_conventions#List_of_x86_calling_conventions

Reply:
Wait-wait-wait. Don't forget that the lowest we ever go is RFT. Isn't RFT's stack struct already ordered the right way?

If not, then the current pascal implementation is broken. On top of that, it's probably not trivial to fix it, because we can't just "reverse" a struct. The endianness starts coming into the play, and we managed to somehow avoid worrying about it until now.

Reply:
Let's take a step back.
The ABI stuff is used for two purposes: 1) going from a `struct` representing stack arguments attached to an RFT to plain arguments in CABI and 2) given some CABI stack arguments get their offset on the stack.

In both cases, the order in which arguments are pushed on the stack is very relevant.

If you an RFT with stack arguments type `struct { uint32_t a; uint32_t b; }` you need to know if it's `void myfunc(uint32_t a, uint32_t b)` or `void myfunc(uint32_t b, uint32_t a)`.

In the same way, for use case 2), given `void myfunc(uint32_t a, uint32_t b)` you need to know if `a` is at offset -4 and `b` at offset `-8` or viceversa.

> we can't just "reverse" a struct

I don't think we need to do that. For 1), we can keep doing what we're doing but simply reverse stack arguments and for 2) we need walk CABI stack arguments backwards, I guess.
But this is a bug if we're not doing this.